pkg/controllermanager/controller/bastion/bastion_control.go

```diff
@@ -0,0 +1,107 @@
+// limitations under the License.
+
+package bastion
+
+import (
+	"context"
+	"time"
+
+	operationsv1alpha1 "github.com/gardener/gardener/pkg/apis/operations/v1alpha1"
+	"github.com/gardener/gardener/pkg/logger"
+
+	"github.com/sirupsen/logrus"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+func (c *Controller) bastionAdd(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		logger.Logger.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+	c.bastionQueue.Add(key)
+}
+
+func (c *Controller) bastionUpdate(_, newObj interface{}) {
+	c.bastionAdd(newObj)
+}
+
+func (c *Controller) bastionDelete(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		logger.Logger.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+	c.bastionQueue.Add(key)
+}
+
+// NewBastionReconciler creates a new instance of a reconciler which reconciles Bastions.
+func NewBastionReconciler(
+	logger logrus.FieldLogger,
+	gardenClient client.Client,
+	maxLifetime time.Duration,
+) reconcile.Reconciler {
+	return &reconciler{
+		logger:       logger,
+		gardenClient: gardenClient,
+		maxLifetime:  maxLifetime,
+	}
+}
+
+type reconciler struct {
+	logger       logrus.FieldLogger
+	gardenClient client.Client
+	maxLifetime  time.Duration
+}
+
+// Reconcile reacts to updates on Bastion resources and cleans up expired Bastions.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	logger := r.logger.WithField("bastion", request)
+
+	bastion := &operationsv1alpha1.Bastion{}
+	if err := r.gardenClient.Get(ctx, request.NamespacedName, bastion); err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("Object is gone, stop reconciling")
+			return reconcile.Result{}, nil
+		}
+
+		logger.Infof("Unable to retrieve object from store: %v", err)
+		return reconcile.Result{}, err
+	}
+
+	// do not reconcile anymore once the object is marked for deletion
+	if bastion.DeletionTimestamp != nil {
+		return reconcile.Result{}, nil
+	}
+
+	// delete the bastion once it has expired
+	if bastion.Status.ExpirationTimestamp != nil && time.Now().After(bastion.Status.ExpirationTimestamp.Time) {
+		logger.WithField("expired", bastion.Status.ExpirationTimestamp.Time).Info("Deleting expired bastion")
+		err := r.gardenClient.Delete(ctx, bastion)
+		return reconcile.Result{}, err
+	}
+
+	// delete the bastion once it has reached its maximum lifetime
+	if time.Since(bastion.CreationTimestamp.Time) > r.maxLifetime {
+		logger.WithField("created", bastion.CreationTimestamp.Time).Info("Deleting bastion because it reached its maximum lifetime")
+		err := r.gardenClient.Delete(ctx, bastion)
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
```

Comment:
Shouldn't we auto-requeue here with the duration of the time until `min(.status.expirationTimestamp, .metadata.creationTimestamp+r.maxLifetime)` so that the `Bastion` gets deleted as soon as one of the deletion condition is met?

Reply:
Can we add like 3 seconds on top of that, to prevent any timing skews? I don't feel too comfortable re-queueing with millisecond precision. ^^

Reply:
Nevermind, it should be fine.